docs/architecture/adr-d001-inter-validator-set-messaging.md

```diff
@@ -0,0 +1,222 @@
+   5. [What are the efficiency considerations (time/space)?](#what-are-the-efficiency-considerations-timespace)
+   6. [What are the expected access patterns (load/throughput)?](#what-are-the-expected-access-patterns-loadthroughput)
+   7. [Are there any logging, monitoring, or observability needs?](#are-there-any-logging-monitoring-or-observability-needs)
+   8. [Are there any security considerations?](#are-there-any-security-considerations)
+   9. [Are there any privacy considerations?](#are-there-any-privacy-considerations)
+   10. [How will the changes be tested?](#how-will-the-changes-be-tested)
+   11. [How will the changes be broken up for ease of review?](#how-will-the-changes-be-broken-up-for-ease-of-review)
+   12. [Will these changes require a breaking (major) release?](#will-these-changes-require-a-breaking-major-release)
+   13. [Does this change require coordination with the SDK or any other software?](#does-this-change-require-coordination-with-the-sdk-or-any-other-software)
+5. [Status](#status)
+6. [Consequences](#consequences)
+   1. [Positive](#positive)
+   2. [Negative](#negative)
+   3. [Neutral](#neutral)
+7. [References](#references)
+
+## Context
+
+### Definitions
+
+For the needs of this document, we define the following types of nodes:
+
+* Validator nodes, which participate in the consensus protocol, verify and sign blocks, 
+* Full nodes, which execute valid blocks.
+
+Note: there are also light nodes, but light nodes are not relevant for the discussion.
+
+Validator Set is a group of Validator nodes responsible for the consensus at a given time. Only one Validator Set can
+be active at the given time. After a predefined number of blocks, the ABCI app initiates the Validator Set rotation to
+make another Validator Set active. 
+
+### Problem statement
+
+The consensus process requires direct communication between validators, as full nodes do not accept nor propagate
+consensus protocol messages (like votes).
+
+Every node, regardless of its type, selects peer nodes they connect to on a random basis. There is no differentiation
+between Full and Validator nodes, and there is no method to ensure that each Validator is (directly or through other
+Validators) connected to all other Validators. In an extreme case, a Validator node can be connected only to Full
+nodes, effectively blocking its communication and excluding it from participation in the consensus.
+
+### Solution
+
+Each Validator node shall allocate a preconfigured number of connectivity slots for communication with other Validators
+that belong to the same Validator Set.
+
+In the current implementation, the ABCI application manages the rotation of Validator Sets. When rotation is needed, 
+the application sends information about the new Validator Set in response to the `EndBlock` message. 
+
+`EndBlock` message shall be extended by adding a network address of each Validator node. Based on that, each
```

Comment:
I would say [`ValidatorUpdate`](https://docs.tendermint.com/v0.32/spec/abci/abci.html#validatorupdate) needs to be extended, which is part of `EndBlock` response.

Reply:
Oh, you have these details below. Just ignore me then.